hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1052,20 +1054,71 @@ int SaraNcpClient::selectSimCard(ModemState& state) {
+            }
+        } while (++ccidCount < 2);
+        // If failed above i.e., netConf_ is still not valid, look for network settings based on IMSI
+        if (!netConf_.isValid()) {
+            int imsiCount = 0;
+            do {
+                memset(buf, 0, sizeof(buf));
+                auto resp = parser_.sendCommand(UBLOX_CIMI_TIMEOUT, "AT+CIMI");
+                CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
+                const int r = CHECK_PARSER(resp.readResult());
+                if (r == AtResponse::OK) {
+                    netConf_ = networkConfigForImsi(buf, strlen(buf));
+                    break;
+                }
+            } while (++imsiCount < 2);
+        }
+        do {
+            // Set UMNOPROF and UBANDMASK as appropriate based on SIM
             auto respUmnoprof = parser_.sendCommand("AT+UMNOPROF?");
             reset = false;
-            int umnoprof = static_cast<int>(UbloxSaraUmnoprof::NONE);
-            auto r = CHECK_PARSER(respUmnoprof.scanf("+UMNOPROF: %d", &umnoprof));
+            int curProf = static_cast<int>(UbloxSaraUmnoprof::NONE);
+            auto r = CHECK_PARSER(respUmnoprof.scanf("+UMNOPROF: %d", &curProf));
             CHECK_PARSER_OK(respUmnoprof.readResult());
-            if (r == 1 && static_cast<UbloxSaraUmnoprof>(umnoprof) == UbloxSaraUmnoprof::SW_DEFAULT) {
+            // First time setup, or switching between official SIM on wrong profile?
+            if (r == 1 && (static_cast<UbloxSaraUmnoprof>(curProf) == UbloxSaraUmnoprof::SW_DEFAULT ||
+                (netConf_.netProv() == CellularNetworkProvider::TWILIO &&
+                static_cast<UbloxSaraUmnoprof>(curProf) != UbloxSaraUmnoprof::STANDARD_EUROPE) ||
+                (netConf_.netProv() == CellularNetworkProvider::KORE_ATT &&
+                static_cast<UbloxSaraUmnoprof>(curProf) != UbloxSaraUmnoprof::ATT)) )
+            {
+                int newProf = static_cast<int>(UbloxSaraUmnoprof::SIM_SELECT);
+                // TWILIO Super SIM
+                if (netConf_.netProv() == CellularNetworkProvider::TWILIO) {
+                    // _oldFirmwarePresent: u-blox firmware 05.06* and 05.07* does not have
+                    // UMNOPROF=100 available. Default to UMNOPROF=0 in that case.
+                    if (oldFirmwarePresent_) {
+                        if (static_cast<UbloxSaraUmnoprof>(curProf) == UbloxSaraUmnoprof::SW_DEFAULT) {
+                            break;
+                        } else {
+                            newProf = static_cast<int>(UbloxSaraUmnoprof::SW_DEFAULT);
+                        }
+                    } else {
+                        newProf = static_cast<int>(UbloxSaraUmnoprof::STANDARD_EUROPE);
+                    }
+                }
+                // KORE AT&T or 3rd Party SIM
+                else {
+                    // break out of do-while if we're trying to set SIM_SELECT a third time
+                    if (resetCount >= 2) {
```

Comment:
We discussed this one in the other PR that was merged, that we want to error on the side of making sure official SIMs get at least one retry, but 3rd party SIMs can just deal with the extra try even though they likely will end up on default anyway.  In reality, if a 3rd party SIM is going to switch to a supported UMNOPROF (2,3,etc..), having an extra try on SIM_SELECT is good for them as well.

Regarding external though... "external" is different from "3rd party".  I don't consider making decisions about external as a necessary choice, because external SIMs could also be official SIMs.

Reply:
Sure why not :D If the external SIM holder is there and a customer wants to switch to a new SIM we offer and not get a new device... seems like a valid option.  We also need this support in our own testing environments.

Reply:
Then, perhaps, it would be better to break the loop early if we know that it's not an official SIM (i.e. we don't have a provider info for it) and, otherwise, still do all 4 attempts for all official SIMs including Kore?

Reply:
I just still don't get this separation: `KORE AT&T or 3rd Party SIM` -> break early

Reply:
We do not officially support 3rd Party SIMs, so we should not be writing code specifically that enhances them.  If it's good enough for official SIMs, it's good enough 👍

Reply:
So 2 attempts is good enough for Kore, but we need 4 for Twilio, is that what you mean? If so, let's be explicit and maybe add a comment?

Reply:
Twilio will likely never see 4 tries (usually 2 with a new SIM)... it's just there in case some combination of UMNOPROF and UBANDMASK fail.

Reply:
Cool, let's mention this in a comment then :pray: